### PR TITLE
CI improvements

### PR DESCRIPTION
**Pull request by kdarkhan**
* Enable caching of build artifacts
  Should considerably speed up builds, especially if building `fuzz` crate.
* Clean up the job `pulldown-cmark-escape`.
  The existing job was confusingly executing workspace tests from the child dir.
* Enable build of `fuzz` and `dos-fuzzer` on every PR.

### CODE REVIEW

**Comment by Martin1887** [comment]
Thanks, but I'm not sure about that caching. I have to check it carefully before merging this.

**Comment by kdarkhan** [comment]
Sure. If you think caching is a riskier change, I can split it up into a separate PR. Even without caching, the builds are not too slow as can be seen from the build results. The longest job with fuzz build enabled is completed in 2 mins. Since cache is empty at the moment (will be populated after merge to master), it does a clean build.

**Comment by Martin1887** [comment]
I think we can remove the caching at all then, thanks.

**Comment by kdarkhan** [comment]
Sure, commit related to caching is removed now.

**Comment by Martin1887** [comment]
The problem I see now is that the `pulldown-cmark-escape` was not totally redundant, since it is not tested now against the 1.71 Rust version, right?

**Comment by ollpu** [comment]
> The problem I see now is that the `pulldown-cmark-escape` was not totally redundant, since it is not tested now against the 1.71 Rust version, right?

It is tested as part of "Cargo test": `cargo test --verbose` since it is a default member of the workspace.

Slightly confused though, because `bench` is also a default member, and it doesn't build / test on 1.71.1 for me locally.

**Comment by ollpu** [comment]
Ah I see, these are done in `working-directory: pulldown-cmark`

Could we make `bench` not a default member, and then run `cargo test` at the top level for Rust 1.71?

**Comment by kdarkhan** [comment]
Yes, `pulldown-cmark-escape` should be tested since it is part of the default workspace (with `--workspace` param).

> Slightly confused though, because bench is also a default member, and it doesn't build / test on 1.71.1 for me locally

Are you sure it was `bench` that is failing? For me with `1.71.1` it builds successfully. What is failing is `dos-fuzzer`.

> Could we make bench not a default member, and then run cargo test at the top level for Rust 1.71 (remove defaults: working-directory completely)? That way it would also run pulldown-cmark-escape tests

Done.

**Comment by ollpu** [comment]
> Are you sure it was bench that is failing? For me with 1.71.1 it builds successfully. What is failing is dos-fuzzer.

Both depend on the same version of `clap`. `bench` has it as a dev-dependency though, so it's only `cargo test` that doesn't work.

File: .github/workflows/rust.yml [hunk 1]
```diff
@@ -15,32 +15,29 @@ jobs:
...
     steps:
     - uses: actions/checkout@v4
     - name: Install Rust
       run: rustup default ${{ matrix.rust-version }}
     - name: Cargo build
+      if: ${{ matrix.rust-version != 'nightly' }}
       run: cargo build --verbose
+    - name: Cargo build the workspace
+      if: ${{ matrix.rust-version == 'nightly' }}
+      run: cargo build --workspace --verbose
```
[the remark this answers is not in the document]
Comment (kdarkhan) [reply .github/workflows/rust.yml]: `cargo-fuzz` requires `nightly` toolchain and I did not want to build `fuzz` in non-nightly toolchain.
I see that `fuzz` package builds with regular `cargo build --workspace`, so I adjusted the condition to ignore only `1.71.1` version.

**Review by ollpu — APPROVED** [review]

**Pull request merged**